### PR TITLE
feat(LoneScale Node): Add LoneScale node and Trigger node

### PR DESCRIPTION
**Pull request by SanYann**
Github issue / Community forum post (link here to close automatically):

### CODE REVIEW

**Comment by Joffcom** [comment]
Hey @SanYann,

I was just starting to review this and I got part of the way through it and I was wondering if this was finished or still a work in progress?

**Comment by SanYann** [comment]
> Hey @SanYann,
> 
> I was just starting to review this and I got part of the way through it and I was wondering if this was finished or still a work in progress?

Hello, It is finished and tested locally

**Comment by Joffcom** [comment]
Hey @SanYann,

Perfect, I had to double check as there were some things that were a bit odd. It looks like the linter is still showing some errors and there is no support for `continueOnFail`. I have added in a credential test and moved all of the requests to use the same function.

When I was looking at the UI I noticed that adding a company and a contact uses the same fields, Is that expected? Looking at the API docs there might not be a first name field for company. Do you know if there is a way to get hold of test credentials or would signing up for a trial be the best bet?

**Comment by SanYann** [comment]
> Hey @SanYann,
> 
> Perfect, I had to double check as there were some things that were a bit odd. It looks like the linter is still showing some errors and there is no support for `continueOnFail`. I have added in a credential test and moved all of the requests to use the same function.
> 
> When I was looking at the UI I noticed that adding a company and a contact uses the same fields, Is that expected? Looking at the API docs there might not be a first name field for company. Do you know if there is a way to get hold of test credentials or would signing up for a trial be the best bet?

Hello,

Thanks a lot for your improvements.
Company and People uses the same fields for now, you can for exemple create a people attached to a company.
To test credentials, I can create a test account with default password on your email address, you will just have to create a new one on your first connection

**Comment by Joffcom** [comment]
Hey @SanYann,

That could work, There is still a bit more tidying up to do and support for features like continue on fail and paired items but I think another couple of sessions and it should be in a good place.

**Comment by SanYann** [comment]
> continueOnFail

Do you have documentation I could follow or example to implements  continue on fail and paired items?

**Comment by SanYann** [comment]
>

**Comment by Joffcom** [comment]
Hey @SanYann,

PairedItems can be found here: https://docs.n8n.io/integrations/creating-nodes/build/reference/paired-items/

ContinueOnFail I can't see anything for but most nodes will have it at the bottom of the item loop, It will look something like this...

```
const executionData = this.helpers.constructExecutionMetaData(
					this.helpers.returnJsonArray(responseData),
					{ itemData: { item: i } },
				);
				returnData.push(...executionData);
			} catch (error) {
				if (this.continueOnFail()) {
					returnData.push({ error: error.message, json: {} });
					continue;
				}
				throw error;
			}
		}
```

If you use the newer declarative style this sort of thing is taken care of for you as well and it normally means you can build out a node fairly quickly as you don't need to worry about some of the more standard things but both styles are valid.

File: packages/nodes-base/credentials/LoneScaleApi.credentials.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+			typeOptions: { password: true },
+			default: '',
+		},
+	];
+
+	authenticate: IAuthenticateGeneric = {
+		type: 'generic',
+		properties: {
+			headers: {
+				Authorization: '=Bearer {{$credentials.apiKey}}',
```
Comment (Joffcom) [inline packages/nodes-base/credentials/LoneScaleApi.credentials.ts:27]: This header key should be...
```
'X-API-KEY': '={{$credentials.apiKey}}',
```
Using the Authorization header with Bearer leads to an authentication failure

File: packages/nodes-base/nodes/LoneScale/LoneScaleList.node.ts [hunk 2]
```diff
@@ -0,0 +1,426 @@
...
+				name: 'operation',
+				type: 'options',
+				displayOptions: {
+					show: {
+						resource: ['item'],
+					},
+				},
+				options: [
+					{
+						name: 'Add',
```
Comment (Joffcom) [inline packages/nodes-base/nodes/LoneScale/LoneScaleList.node.ts:86]: We prefer Create over Add for operation names

File: packages/nodes-base/nodes/LoneScale/LoneScaleList.node.ts [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+					show: {
+						resource: ['item'],
+					},
+				},
+				options: [
+					{
+						name: 'Company',
+						value: 'COMPANY',
+						description: 'List of company',
+						action: 'Get company lists',
```
Comment (Joffcom) [inline packages/nodes-base/nodes/LoneScale/LoneScaleList.node.ts:110]: This is a create not a get

File: packages/nodes-base/nodes/LoneScale/LoneScaleList.node.ts [hunk 4]
```diff
@@ -0,0 +1,426 @@
...
+					},
+					{
+						name: 'Contact',
+						value: 'PEOPLE',
+						description: 'List of contact',
```
Comment (Joffcom) [inline packages/nodes-base/nodes/LoneScale/LoneScaleList.node.ts:115]: As above

File: packages/nodes-base/nodes/LoneScale/LoneScaleList.node.ts [hunk 5]
```diff
@@ -0,0 +1,426 @@
...
+					loadOptionsMethod: 'getLists',
+					loadOptionsDependsOn: ['type'],
+				},
+				default: '',
+				description:
+					'Choose from the list, or specify an ID using an <a href="https://docs.n8n.io/code-examples/expressions/">expression</a>',
+				required: true,
+			},
+			{
+				displayName: 'First Name',
```
Comment (Joffcom) [inline packages/nodes-base/nodes/LoneScale/LoneScaleList.node.ts:142]: This should be marked as required as it is needed to create a person on a list

File: packages/nodes-base/nodes/LoneScale/LoneScaleList.node.ts [hunk 6]
```diff
@@ -0,0 +1,426 @@
...
+						operation: ['add'],
+						resource: ['item'],
+						type: ['PEOPLE'],
+					},
+				},
+				default: '',
+				description: 'Contact first name',
+			},
+			{
+				displayName: 'Last Name',
```
Comment (Joffcom) [inline packages/nodes-base/nodes/LoneScale/LoneScaleList.node.ts:156]: This should be marked as required as it is needed to create a person on a list

File: packages/nodes-base/nodes/LoneScale/LoneScaleList.node.ts [hunk 7]
```diff
@@ -0,0 +1,426 @@
...
+						operation: ['add'],
+						resource: ['item'],
+						type: ['PEOPLE'],
+					},
+				},
+				default: '',
+				description: 'Contact last name',
+			},
+			{
+				displayName: 'Full Name',
```
Comment (Joffcom) [inline packages/nodes-base/nodes/LoneScale/LoneScaleList.node.ts:170]: This is optional, Should be moved under an Additional Items field

File: packages/nodes-base/nodes/LoneScale/LoneScaleList.node.ts [hunk 8]
```diff
@@ -0,0 +1,426 @@
...
+						operation: ['add'],
+						resource: ['item'],
+						type: ['PEOPLE'],
+					},
+				},
+				default: '',
+				description: 'Contact full name',
+			},
+			{
+				displayName: 'Linkedin Url',
```
Comment (Joffcom) [inline packages/nodes-base/nodes/LoneScale/LoneScaleList.node.ts:184]: This is optional for a person, Should be moved under an Additional Items field

File: packages/nodes-base/nodes/LoneScale/LoneScaleList.node.ts [hunk 9]
```diff
@@ -0,0 +1,426 @@
...
+					show: {
+						operation: ['add'],
+						resource: ['item'],
+					},
+				},
+				default: '',
+				description: 'Contact Linkedin URL',
+			},
+			{
+				displayName: 'Company Name',
```
Comment (Joffcom) [inline packages/nodes-base/nodes/LoneScale/LoneScaleList.node.ts:197]: This is optional for a person, Should be moved under an Additional Items field. For a Company it should be marked as required.

File: packages/nodes-base/nodes/LoneScale/LoneScaleList.node.ts [hunk 10]
```diff
@@ -0,0 +1,426 @@
...
+						operation: ['add'],
+						resource: ['item'],
+						type: ['PEOPLE'],
+					},
+				},
+				default: '',
+				description: 'Contact current position',
+			},
+			{
+				displayName: 'Company Domain',
```
Comment (Joffcom) [inline packages/nodes-base/nodes/LoneScale/LoneScaleList.node.ts:224]: This is optional for both person and company, Should be moved under an Additional Items field

File: packages/nodes-base/nodes/LoneScale/LoneScaleList.node.ts [hunk 11]
```diff
@@ -0,0 +1,426 @@
...
+					show: {
+						operation: ['add'],
+						resource: ['item'],
+					},
+				},
+				default: '',
+				description: 'Contact company domain',
+			},
+			{
+				displayName: 'Contact Location',
```
Comment (Joffcom) [inline packages/nodes-base/nodes/LoneScale/LoneScaleList.node.ts:237]: This is optional for both person and company, Should be moved under an Additional Items field

File: packages/nodes-base/nodes/LoneScale/LoneScaleList.node.ts [hunk 12]
```diff
@@ -0,0 +1,426 @@
...
+				displayOptions: {
+					show: {
+						operation: ['add'],
+						resource: ['item'],
+					},
+				},
+				default: '',
+			},
+			{
+				displayName: 'Contact Email',
```
Comment (Joffcom) [inline packages/nodes-base/nodes/LoneScale/LoneScaleList.node.ts:249]: This is optional, Should be moved under an Additional Items field

File: packages/nodes-base/nodes/LoneScale/LoneScaleList.node.ts [hunk 13]
```diff
@@ -0,0 +1,426 @@
...
+				if (resource === 'list') {
+					if (operation === 'create') {
+						const name = this.getNodeParameter('name', i) as string;
+						const entity = this.getNodeParameter('type', i) as string;
+						const body: IDataObject = {
+							name,
+							entity,
+						};
+
+						console.log({ body });
```
Comment (Joffcom) [inline packages/nodes-base/nodes/LoneScale/LoneScaleList.node.ts:355]: This probably isn't needed and can be removed

**Review by Joffcom — APPROVED** [review]

**Comment by janober** [comment]
Got released with [`n8n@0.230.1`](https://github.com/n8n-io/n8n/releases/tag/n8n%400.230.1)

**Pull request merged**